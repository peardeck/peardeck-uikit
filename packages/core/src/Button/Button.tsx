import styled from "styled-components";
import {
  border,
  BorderProps,
  color,
  ColorProps,
  layout,
  LayoutProps,
  shadow,
  ShadowProps,
  space,
  SpaceProps,
  typography,
  TypographyProps,
  variant,
} from "styled-system";
import { themeGet } from "@styled-system/theme-get";
import { darken } from "polished";
import theme, { ThemedComponentProps } from "../Theme/theme";

export type ButtonVariant = "default" | "primary" | "secondary";

export type ButtonSize = "small" | "medium" | "large";

export type ButtonProps = BorderProps &
  ColorProps &
  LayoutProps &
  ShadowProps &
  SpaceProps &
  ThemedComponentProps &
  TypographyProps & {
    disabled?: boolean;
    buttonSize?: ButtonSize;
    variant?: ButtonVariant;
  };

// TODO: do not hardcode background hover color.

export const Button: React.FunctionComponent<ButtonProps> = styled.button<
  ButtonProps
>`
  cursor: ${({ disabled }): string => (disabled ? "not-allowed" : "pointer")};
  transition: all 0.2s ease;

  > * {
    vertical-align: middle;
  }

  :disabled {
    opacity: 0.3;
  }

  /* By default calculate a darkened hover color -- but this can be overriden by the variant */
  :hover {
    background-color: ${(props): string =>
      darken(0.1, themeGet(`colors.${props.variant}`, "#ffffff")(props))}}
  }

  ${variant({
    prop: "variant",
    variants: {
      default: {
        color: "text",
        bg: "background",
      },
      primary: {
        color: "white",
        bg: "primary",
      },
      secondary: {
        color: "white",
        bg: "secondary",
      },
    },
  })}

  ${variant({
    prop: "buttonSize",
    variants: {
      small: {
        fontSize: 0,
        px: 2,
        py: 1,
      },
      medium: {
        fontSize: 2,
        px: 3,
        py: 2,
      },
      large: {
        fontSize: 4,
        px: 4,
        py: 3,
      },
    },
  })}

  ${layout}
  ${border}
  ${color}
  ${shadow}
  ${space}
  ${typography}
`;

Button.defaultProps = {
  borderRadius: 3,
  boxShadow: "medium",
  buttonSize: "small",
  fontFamily: "inherit",
  fontWeight: "bold",
  lineHeight: "default",
  textAlign: "center",
  theme,
  variant: "default",
};

Button.displayName = "Button";
