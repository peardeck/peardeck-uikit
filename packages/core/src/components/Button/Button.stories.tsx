import React, { ElementType } from "react";
import { storiesOf } from "@storybook/react";
import { select, text, boolean } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";
import {
  IconAlphabetical,
  IconAccount,
  IconCalendar,
  IconCheckmark,
  IconClassroom,
} from "@peardeck-uikit/icons";
import { Button, ButtonSize, ButtonVariant } from "./Button";
import { themeSpaceKnob } from "../../theme/knobs.stories";

const disabledKnob = (initial = false) => boolean("Disabled", initial);

const buttonSizeKnob = (initial: ButtonSize = "medium") =>
  select("Size", ["small", "medium", "large"], initial) as ButtonSize;

const buttonVariantKnbo = (initial: ButtonVariant = "default") =>
  select(
    "Variant",
    ["default", "primary", "secondary"],
    initial
  ) as ButtonVariant;

const ICONS: { [key: string]: ElementType } = {
  IconAlphabetical,
  IconAccount,
  IconCalendar,
  IconCheckmark,
  IconClassroom,
};
const ICONS_KEYS = Object.keys(ICONS);

storiesOf("core/Button", module)
  .add(
    "with text",
    () => (
      <Button
        onClick={action("onClick")}
        disabled={disabledKnob()}
        buttonSize={buttonSizeKnob()}
        m={themeSpaceKnob("Margin")}
        p={themeSpaceKnob("Padding")}
        variant={buttonVariantKnbo()}
      >
        {text("Label", "Hello!")}
      </Button>
    ),
    {
      notes: `
        # Button

        Button is used for actions, like in forms. Use \`Link\` components when navigating a page or changing the URL.

        In special cases where you'd like to use a <a> styled like a Button, use <Button as='a'> and provide an href.
      `,
    }
  )
  .add("with icon", () => {
    const iconKey = select("Icon", ICONS_KEYS, ICONS_KEYS[0]);
    const IconComponent = ICONS[iconKey];

    return (
      <Button
        onClick={action("onClick")}
        disabled={disabledKnob()}
        buttonSize={buttonSizeKnob()}
        variant={buttonVariantKnbo()}
      >
        <IconComponent /> {text("Label", "This button has an icon!")}
      </Button>
    );
  })
  .add("with emoji", () => (
    <Button
      onClick={action("onClick")}
      disabled={disabledKnob()}
      buttonSize={buttonSizeKnob()}
      variant={buttonVariantKnbo()}
    >
      <span role="img" aria-label="so cool">
        {text("Label", "😀 😎 👍 💯")}
      </span>
    </Button>
  ));
