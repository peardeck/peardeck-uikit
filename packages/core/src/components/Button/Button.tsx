import styled from "styled-components";
import {
  border,
  BorderProps,
  color,
  ColorProps,
  layout,
  LayoutProps,
  shadow,
  ShadowProps,
  space,
  SpaceProps,
  typography,
  TypographyProps,
  variant,
} from "styled-system";
import { themeGet } from "@styled-system/theme-get";
import { darken } from "polished";
import { theme, ThemedComponentProps } from "../../theme/theme";

export type ButtonVariant = "default" | "primary" | "secondary";

export type ButtonSize = "small" | "medium" | "large";

export type ButtonProps = BorderProps &
  ColorProps &
  LayoutProps &
  ShadowProps &
  SpaceProps &
  ThemedComponentProps &
  TypographyProps & {
    disabled?: boolean;
    buttonSize?: ButtonSize;
    variant?: ButtonVariant;
  };

// TODO: fix icon color for darker backgrounds

/**
 * Button is used for actions, like in forms. Use \`Link\` components when
 * navigating a page or changing the URL.
 *
 * In special cases where you'd like to use a <a> styled like a Button,
 * use <Button as='a'> and provide an href.
 */
export const Button = styled.button<ButtonProps>`
  cursor: ${({ disabled }): string => (disabled ? "not-allowed" : "pointer")};
  transition: all 0.2s ease;

  > * {
    vertical-align: middle;
  }

  :disabled {
    opacity: 0.3;
  }

  /* By default calculate a darkened hover color -- but this can be overriden by the variant */
  :hover {
    background-color: ${(props): string =>
      darken(0.1, themeGet(`colors.${props.variant}`, "#ffffff")(props))}}
  }

  ${variant({
    prop: "variant",
    variants: {
      default: {
        color: "text",
        bg: "background",
      },
      primary: {
        color: "white",
        bg: "primary",
      },
      secondary: {
        color: "white",
        bg: "secondary",
      },
    },
  })}

  ${variant({
    prop: "buttonSize",
    variants: {
      small: {
        fontSize: 0,
        px: 2,
        py: 1,
      },
      medium: {
        fontSize: 2,
        px: 3,
        py: 2,
      },
      large: {
        fontSize: 4,
        px: 4,
        py: 3,
      },
    },
  })}

  ${layout}
  ${border}
  ${color}
  ${shadow}
  ${space}
  ${typography}
`;

Button.defaultProps = {
  borderRadius: 3,
  boxShadow: 2,
  buttonSize: "medium",
  fontFamily: "inherit",
  fontWeight: "bold",
  lineHeight: "default",
  textAlign: "center",
  theme,
  variant: "default",
};

Button.displayName = "Button";
