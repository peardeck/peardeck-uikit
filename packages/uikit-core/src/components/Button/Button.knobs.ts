import { select, boolean } from "@storybook/addon-knobs";
import {
  ButtonColorVariant,
  ButtonSizeVariant,
  ButtonStyleVariant,
  SIZE_VARIANTS,
} from "./Button";
import { theme } from "../../theme/theme";

export const disabledKnob = (initial = false) => boolean("Disabled", initial);

export const buttonSizeVariantKnob = (initial: ButtonSizeVariant = "medium") =>
  select("Size Variant", SIZE_VARIANTS, initial) as ButtonSizeVariant;

export const buttonColorVariantKnob = (
  initial: ButtonColorVariant = "default"
) =>
  select(
    "Color Variant",
    Object.keys(theme.buttonColors),
    initial
  ) as ButtonColorVariant;

export const buttonStyleVariantKnob = (
  initial: ButtonStyleVariant = "default"
) =>
  select(
    "Style Variant",
    Object.keys(theme.buttonStyles),
    initial
  ) as ButtonStyleVariant;
