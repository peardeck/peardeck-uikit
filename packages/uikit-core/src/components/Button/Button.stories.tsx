import React, { ElementType } from "react";
import { select, text, boolean } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";
import {
  IconAlphabetical,
  IconAccount,
  IconCalendar,
  IconCheckmark,
  IconClassroom,
} from "@peardeck/uikit-icons";
import { Button, ButtonSize, ButtonColorVariant } from "./Button";
import { themeSpaceKnob } from "../../theme/knobs";
import { theme } from "../../theme/theme";

const disabledKnob = (initial = false) => boolean("Disabled", initial);

const buttonSizeKnob = (initial: ButtonSize = "medium") =>
  select("Size", ["small", "medium", "large"], initial) as ButtonSize;

const buttonColorVariantKnob = (initial: ButtonColorVariant = "default") =>
  select(
    "Color Variant",
    Object.keys(theme.buttonColors),
    initial
  ) as ButtonColorVariant;

const ICONS: { [key: string]: ElementType } = {
  IconAlphabetical,
  IconAccount,
  IconCalendar,
  IconCheckmark,
  IconClassroom,
};
const ICONS_KEYS = Object.keys(ICONS);

export default {
  component: Button,
  title: "core|Button",
};

export const Default = () => (
  <Button
    onClick={action("onClick")}
    disabled={disabledKnob()}
    buttonSize={buttonSizeKnob()}
    m={themeSpaceKnob("Margin")}
    p={themeSpaceKnob("Padding")}
    colorVariant={buttonColorVariantKnob()}
  >
    {text("Label", "Hello!")}
  </Button>
);

export const WithIcon = () => {
  const iconKey = select("Icon", ICONS_KEYS, ICONS_KEYS[0]);
  const IconComponent = ICONS[iconKey];

  return (
    <Button
      onClick={action("onClick")}
      disabled={disabledKnob()}
      buttonSize={buttonSizeKnob()}
      colorVariant={buttonColorVariantKnob()}
    >
      <IconComponent /> {text("Label", "This button has an icon!")}
    </Button>
  );
};
export const WithEmoji = () => (
  <Button
    onClick={action("onClick")}
    disabled={disabledKnob()}
    buttonSize={buttonSizeKnob()}
    colorVariant={buttonColorVariantKnob()}
  >
    <span role="img" aria-label="so cool">
      {text("Label", "😀 😎 👍 💯")}
    </span>
  </Button>
);
