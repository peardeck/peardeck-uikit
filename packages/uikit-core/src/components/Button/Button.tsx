import styled from "styled-components";
import {
  border,
  BorderProps,
  color,
  ColorProps,
  layout,
  LayoutProps,
  shadow,
  ShadowProps,
  space,
  SpaceProps,
  typography,
  TypographyProps,
  variant,
  system,
} from "styled-system";
import { theme, ThemedComponentProps } from "../../theme/theme";

export type ButtonColorVariant = keyof typeof theme.buttonColors; // "default" | "primary" | "secondary";

export type ButtonSize = "small" | "medium" | "large";

export type ButtonProps = BorderProps &
  ColorProps &
  LayoutProps &
  ShadowProps &
  SpaceProps &
  ThemedComponentProps &
  TypographyProps & {
    disabled?: boolean;
    buttonSize?: ButtonSize;
    colorVariant?: ButtonColorVariant;
  };

// TODO: fix icon color for darker backgrounds

/**
 * Button is used for actions, like in forms. Use \`Link\` components when
 * navigating a page or changing the URL.
 *
 * In special cases where you'd like to use a <a> styled like a Button,
 * use <Button as='a'> and provide an href.
 */
export const Button = styled.button.attrs<ButtonProps>(
  ({ theme: currentTheme, colorVariant }) => ({
    hoverBoxShadow: "buttonHover",
    activeBoxShadow: "buttonActive",
    boxShadow: "button",
    ...currentTheme.buttonColors[colorVariant || "default"],
  })
)<ButtonProps>`
  cursor: ${({ disabled }): string => (disabled ? "not-allowed" : "pointer")};
  transition: all 0.2s ease;

  > * {
    vertical-align: middle;
  }

  :disabled {
    opacity: 0.3;
  }

  ${variant({
    prop: "buttonSize",
    variants: {
      small: {
        fontSize: 0,
        px: 2,
        py: 1,
      },
      medium: {
        fontSize: 2,
        px: 3,
        py: 2,
      },
      large: {
        fontSize: 4,
        px: 4,
        py: 3,
      },
    },
  })}

  :hover {
    ${system({
      hoverBackgroundColor: {
        property: "backgroundColor",
        scale: "colors",
      },
      hoverColor: {
        property: "color",
        scale: "colors",
      },
      hoverBoxShadow: {
        property: "boxShadow",
        scale: "shadows",
      },
    })}
  }

  :active {
    ${system({
      activeBgColor: {
        property: "backgroundColor",
        scale: "colors",
      },
      activeColor: {
        property: "color",
        scale: "colors",
      },
      activeBoxShadow: {
        property: "boxShadow",
        scale: "shadows",
      },
    })}
  }
  ${layout}
  ${border}
  ${color}
  ${shadow}
  ${space}
  ${typography}
`;

Button.defaultProps = {
  borderRadius: 3,
  borderWidth: 2,
  boxShadow: 2,
  buttonSize: "medium",
  fontFamily: "inherit",
  fontWeight: "bold",
  lineHeight: "default",
  textAlign: "center",
  theme,
  colorVariant: "default",
};

Button.displayName = "Button";
