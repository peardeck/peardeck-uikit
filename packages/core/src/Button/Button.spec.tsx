import React from "react";
import { create } from "react-test-renderer";
import { Button } from "./Button";

describe("Button", () => {
  test("is a named export", () => {
    expect(Button).toBeDefined();
  });

  test("it matches the snapshot", () => {
    const component = create(<Button />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
