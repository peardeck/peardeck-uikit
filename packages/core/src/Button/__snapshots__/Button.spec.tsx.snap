// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button it matches the snapshot 1`] = `
.c0 {
  border-radius: 10px;
  width: medium;
  height: medium;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.27);
  font-family: inherit;
  font-weight: 700;
  line-height: 1.5;
  text-align: center;
  cursor: pointer;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
  font-size: 1em;
  padding: 0.5em 1em;
  border: 1px solid #dce2e9;
  color: #354242;
  background-color: #ffffff;
}

.c0:hover {
  background-color: rgba(237,240,244,0.42);
}

.c0 > * {
  vertical-align: middle;
}

.c0:disabled {
  opacity: 0.3;
}

<button
  className="c0"
  fontFamily="inherit"
  fontWeight="bold"
  size="medium"
/>
`;
